lib/ansible/modules/crypto/openssl_cert.py

```diff
@@ -0,0 +1,446 @@
+        self.challenge_path = module.params['letsencrypt_challenge']
+
+    def generate(self):
+        if not self.force and os.path.exists(self.dest):
+            self.changed = False
+            return
+
+        backupdest = ""
+        if self.changed and not self.module.check_mode:
+            if self.backup and os.path.exists(self.dest):
+                backupdest = module.backup_local(self.dest)
+
+            # TODO (spredzy): Ugly part should be done directly by interacting
+            # with the acme protocol through python-acme
+            try:
+                os.system('acme-tiny --account-key %s --csr %s --acme-dir %s > %s' % (self.accountkey, self.csr_path, self.challenge_path, self.dest))
+            except:
+                e = get_exception()
+                raise CertificateError(e)
+
+    def dump(self):
+
+        result = {
+            'changed': self.changed,
+            'cert_path': self.dest,
+            'account_key': self.accountkey,
+            'backup': backupdest,
+        }
+
+        return result
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            dest=dict(required=True, aliases=['path', 'destfile'], type='path'),
+            provider=dict(required=True, choices=['selfsigned', 'assertonly', 'letsencrypt'], type='str'),
+            force=dict(default=False, type='bool'),
+            backup=dict(default=False, type='bool'),
+            csr=dict(type='str'),
+
+            # General properties of a certificate
+            privatekey=dict(type='path'),
+            signature_algorithms=dict(type='list'),
+            issuer=dict(type='dict'),
+            subject=dict(type='dict'),
+            version=dict(type='int'),
+
+            # provider: selfsigned
+            selfsigned_digest=dict(default='sha256', type='str'),
+            selfsigned_notBefore=dict(default=0, type='int'),
+            selfsigned_notAfter=dict(default=315360000, type='int'),
+
+            # provider: letsencrypt
+            letsencrypt_accountkey=dict(type='str'),
+            letsencrypt_challenge_path=dict(type='path'),
+        )
+    )
+
+    if not pyopenssl_found:
+        module.fail_json(msg='the python pyOpenSSL module is required')
+
+    provider = module.params['provider']
+    if provider == 'selfsigned':
+        if module.params['csr'] is None:
+            module.fail_json(msg='No csr provided')
+        cert = SelfSignedCertificate(module)
+    elif provider == 'letsencrypt':
+        if module.params['csr'] is None:
+            module.fail_json(msg='No csr provided')
+        cert = LetsEncryptCertificate(module)
+    else:
+        cert = AssertOnlyCertificate(module)
+
+    if module.params['state'] == 'present':
+        try:
+            cert.generate()
+        except CertificateError:
+            module.fail_json(msg='An error occured while generating the certificate')
```

Comment:
Not in this state, since this catches any general exceptions from both providers when generating certificates.